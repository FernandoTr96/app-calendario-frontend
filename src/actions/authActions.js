import { fetchWithoutToken, fetchWithToken } from "../helpers/fetch";
import { types } from "../types/types";
import { toast } from 'react-toastify';

export const startLoginWithEmailAndPassword = (email, password)=>{

    return async (dispatch)=>{

        const response = await fetchWithoutToken(
            'auth',
            {email, password},
            'POST'
        );

        const body = await response.json();
        
        if(body.ok){
            localStorage.setItem('calendarAppToken', body.token );
            localStorage.setItem('token-init-date', new Date().getTime() );
            delete body.token;
            delete body.ok;
            dispatch(login(body));
        }
        else{

            toast.error('usuario o contraseÃ±a incorrecta', {
                position: "top-right",
                autoClose: 5000,
                hideProgressBar: false,
                closeOnClick: true,
                pauseOnHover: false,
                draggable: false,
                progress: undefined,
            });

        }

    }

}

export const login = (userCredentials)=>({
    type: types.login,
    payload: userCredentials
})

export const registerAccount = (userdata)=>{
    
    return async (dispatch)=>{

        const response = await fetchWithoutToken(
            'auth/register',
            {...userdata},
            'POST'
        );

        const body = await response.json();

        if(body.ok){
            localStorage.setItem('calendarAppToken', body.token );
            localStorage.setItem('token-init-date', new Date().getTime() );
            delete body.token;
            delete body.ok;
            delete body.confirm_password;
            dispatch(login(body));
        }
        else{

            toast.error(body.msg, {
                position: "top-right",
                autoClose: 5000,
                hideProgressBar: false,
                closeOnClick: true,
                pauseOnHover: false,
                draggable: false,
                progress: undefined,
            });

        }

        if(body.errors?.length > 0){
            body.errors.forEach(field => {

                toast.error(field.msg, {
                    position: "top-right",
                    autoClose: 5000,
                    hideProgressBar: false,
                    closeOnClick: true,
                    pauseOnHover: false,
                    draggable: false,
                    progress: undefined,
                });

            })
        }

    }

}

export const startChecking = ()=> {

    return async (dispatch)=>{

        const response = await fetchWithToken(
            'auth/revalidate-token',
            'GET'
        );

        const body = await response.json();

        if(body.ok){
            localStorage.setItem('calendarAppToken', body.token );
            localStorage.setItem('token-init-date', new Date().getTime() );
            delete body.token;
            delete body.ok;
            dispatch(login(body));
        }
        else{
            dispatch(checkingFinish());
        }

    }

}

export const checkingFinish = ()=>({
    type: types.checkingFinish
})

export const startLogout = ()=>{

    return (dispatch)=>{

        localStorage.clear();
        dispatch(logout());

    }

}

export const logout = ()=>({
    type: types.purgeAuth
})