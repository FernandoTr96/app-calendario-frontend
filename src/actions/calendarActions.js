import { types } from "../types/types";
import {fetchWithToken} from '../helpers/fetch'
import {toast} from 'react-toastify'
import { string2date } from "../helpers/string2date";

export const getEvents = ()=>{
    return async (dispatch)=>{

        dispatch(showLoader());

        try {

            const response = await fetchWithToken('events');
            const body = await response.json();
            

            if(response.ok){
                
                const events = string2date(body.events);
                dispatch(setEvents(events));

            }

            
        } catch (error) {

            toast.error(`Error : ${error}`, {
                position: "bottom-right",
                autoClose: 4000,
                hideProgressBar: false,
                closeOnClick: true,
                pauseOnHover: false,
                draggable: false,
                progress: undefined,
            });

        }

        dispatch(hiddeLoader());

    }
}

export const setEvents = (events) =>
(
    {
        type: types.fillEventList,
        payload: events
    }
)

export const saveEvent = (event)=>{
    return async (dispatch) => {

        dispatch(showLoader());

        try {

            const insert = await fetchWithToken(
                'events/addEvent',
                event,
                'POST'
            );

            if(insert.ok){
                toast.success(`Se ha aÃ±adido un nuevo evento al calendario`, {
                    position: "bottom-right",
                    autoClose: 3000,
                    hideProgressBar: false,
                    closeOnClick: true,
                    pauseOnHover: false,
                    draggable: false,
                    progress: undefined,
                });
                dispatch(getEvents());
            }
            
        } catch (error) {
            
            toast.error(error.msg, {
                position: "bottom-right",
                autoClose: 5000,
                hideProgressBar: false,
                closeOnClick: true,
                pauseOnHover: false,
                draggable: false,
                progress: undefined,
            });

        }

        dispatch(hiddeLoader());

    }
}


export const editEvent = (event)=>{
    return async (dispatch)=>{

        const id = event.id;
        delete event.id;

        dispatch(showLoader());

        try {

            const update =  await fetchWithToken(
                `events/update/${id}`,
                event,
                'PUT'
            );

            if(update.ok){

                toast.info(`evento actualizado con exito`, {
                    position: "bottom-right",
                    autoClose: 3000,
                    hideProgressBar: false,
                    closeOnClick: true,
                    pauseOnHover: false,
                    draggable: false,
                    progress: undefined,
                });

                dispatch(getEvents());

            }

        } catch (error) {
            
            toast.error(`Error: ${error.msg}`, {
                position: "bottom-right",
                autoClose: 5000,
                hideProgressBar: false,
                closeOnClick: true,
                pauseOnHover: false,
                draggable: false,
                progress: undefined,
            });
        }

        dispatch(hiddeLoader());

    }
}

export const deleteEvent = (id)=>{
    return async (dispatch)=>{

        dispatch(showLoader());

        try {

            const _delete = await fetchWithToken(
                `events/delete/${id}`,
                {},
                'DELETE'
            )

            if(_delete.ok){

                toast.warning(`el evento fue eliminado con exito`, {
                    position: "bottom-right",
                    autoClose: 3000,
                    hideProgressBar: false,
                    closeOnClick: true,
                    pauseOnHover: false,
                    draggable: false,
                    progress: undefined,
                });

                dispatch(getEvents());
            }

            
        } catch (error) {

            toast.error(`Error: ${error.msg}`, {
                position: "bottom-right",
                autoClose: 5000,
                hideProgressBar: false,
                closeOnClick: true,
                pauseOnHover: false,
                draggable: false,
                progress: undefined,
            });

        }

        dispatch(hiddeLoader());

    }
}

export const activeEvent = (event)=>(
    {
        type: types.activeEvent,
        payload: event
    }
)

export const showLoader = ()=>(
    {
        type: types.showLoader
    }
)
export const hiddeLoader = ()=>(
    {
        type: types.hiddeLoader
    }
)