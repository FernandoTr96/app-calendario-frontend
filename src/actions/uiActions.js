import { types } from "../types/types";

export const openModalEvents = ()=>(
    {
        type: types.uiOpenModalEvents,
        payload: true
    }
);

export const closeModalEvents = ()=>(
    {
        type: types.uiCloseModalEvents,
        payload: false
    }
);