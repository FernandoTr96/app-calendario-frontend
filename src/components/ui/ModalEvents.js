import { Modal, Box, Typography, TextField, Button } from '@mui/material'
import styles from '../../styles/mui-sx-styles/ModalEventsStyles'
import { useForm } from 'react-hook-form'
import moment from 'moment'
import { useDispatch, useSelector } from 'react-redux'
import { closeModalEvents } from '../../actions/uiActions'
import { editEvent, saveEvent, activeEvent } from '../../actions/calendarActions'
import { useEffect } from 'react'

export const ModalEvents = () => {

    const dispatch = useDispatch();
    const { modalEventsOpen } = useSelector(state => state.ui);
    const { activeNow } = useSelector(state => state.events);

    const { register, handleSubmit, watch, reset, getValues, setValue, formState: { errors } } = useForm({
        mode: 'onSubmit'
    });

    const submit = () => {

        const formValues = getValues();
        const newEvent = {
            title: formValues.title,
            desc: formValues.desc,
            start: moment(formValues.startDate),
            end: moment(formValues.endDate)
        }

        if (activeNow === null) {

            dispatch(saveEvent(newEvent));
            reset();
            closeModal();
        }
        else {
            dispatch(editEvent({
                ...newEvent,
                id: activeNow.id
            }));
            reset();
            closeModal();
            dispatch(activeEvent(null));
        }

    }

    const closeModal = () => {
        dispatch(closeModalEvents());
    }

    useEffect(() => {

        if(activeNow !== null)
        {
            setValue('startDate', moment(activeNow.start).format('YYYY-MM-DDThh:mm'));
            setValue('endDate', moment(activeNow.end).format('YYYY-MM-DDThh:mm'));
            setValue('title', activeNow.title);
            setValue('desc', activeNow.desc);
        }
        else
        {
            reset();
        }

    }, [activeNow,setValue,reset])

    return (
        <Modal
            open={modalEventsOpen}
            onClose={closeModal}
            aria-labelledby="modal-modal-title"
            aria-describedby="modal-modal-description"
        >
            <Box onSubmit={handleSubmit(submit)} sx={styles.modal} component='form' autoComplete='off'>
                <Typography variant="h5" component="h5">
                    {activeNow === null ? 'Nuevo evento' : 'Editar evento'}
                </Typography>

                <TextField
                    {...register('startDate', {
                        validate: {
                            isAfter: v => moment(v).isAfter(watch('endDate')) === false,
                            isSame: v => moment(v).isSame(watch('endDate')) === false
                        }
                    })}
                    label="Fecha y hora de inicio"
                    type="datetime-local"
                    fullWidth
                    helperText={
                        errors.startDate?.type === 'isAfter' ? 'La fecha inicial no puede ser despues a la fecha final' :
                            errors.startDate?.type === 'isSame' && 'La fecha inicial no puede ser igual a la fecha final'
                    }
                    error={errors.startDate?.type ? true : false}
                    defaultValue={moment().format('YYYY-MM-DDT12:00')}
                />

                <TextField
                    {...register('endDate', {
                        validate: {
                            isBefore: v => moment(v).isBefore(watch('startDate')) === false,
                            isSame: v => moment(v).isSame(watch('startDate')) === false
                        }
                    })}
                    label="Fecha y hora final"
                    type="datetime-local"
                    fullWidth
                    helperText={
                        errors.endDate?.type === 'isBefore' ? 'La fecha final no puede ser antes a la fecha inicial' :
                            errors.endDate?.type === 'isSame' && 'La fecha final no puede ser igual a la fecha inicial'
                    }
                    error={errors.endDate?.type ? true : false}
                    defaultValue={moment().add(1, 'days').format('YYYY-MM-DDT00:00')}
                />

                <TextField
                    {...register('title', {
                        required: true,
                        pattern: /^[A-Za-z0-9áéíóúÁÉÍÓÚñÑ\s\n\t]{0,50}$/
                    })}
                    label="Titulo"
                    variant="outlined"
                    fullWidth
                    helperText={
                        errors.title?.type === 'required' ? 'Este campo es requerido' :
                        errors.title?.type === 'pattern' && `Debe ser menor a 50 caracteres y sin simbolos extraños | caracteres (${watch('title').length})`}
                    error={errors.title?.type ? true : false}
                />

                <TextField
                    {...register('desc', {
                        required: true,
                        pattern: /^[A-Za-z0-9áéíóúÁÉÍÓÚñÑ\s\n\t]{0,100}$/
                    })}
                    label="Descripcion"
                    multiline
                    rows={4}
                    fullWidth
                    helperText={
                        errors.desc?.type === 'required' ? 'Este campo es requerido' :
                        errors.desc?.type === 'pattern' && `Debe ser menor a (100) caracteres y sin simbolos extraños | caracteres (${watch('desc').length})`
                    }
                    error={errors.desc?.type ? true : false}
                />

                <Button type='submit' variant='gradient-citrus' fullWidth>
                    {activeNow === null ? 'Guardar' : 'Editar'}
                </Button>

            </Box>

        </Modal>
    )
}
