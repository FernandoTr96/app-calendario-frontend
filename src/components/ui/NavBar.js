import React from 'react'
import { Stack, Avatar, Tooltip, AppBar, Toolbar, Typography, IconButton, Box } from '@mui/material'
import LogoutIcon from '@mui/icons-material/Logout';
import EventIcon from '@mui/icons-material/Event';
import DeleteForeverIcon from '@mui/icons-material/DeleteForever';
import EditIcon from '@mui/icons-material/Edit';
import styles from '../../styles/mui-sx-styles/NavBarStyles'
import { stringAvatar } from '../../helpers/mui-avatar-letters';
import { useDispatch, useSelector } from 'react-redux'
import { openModalEvents } from '../../actions/uiActions';
import { activeEvent, deleteEvent } from '../../actions/calendarActions';
import { startLogout } from '../../actions/authActions';


export const NavBar = () => {

    const dispatch = useDispatch();
    const activeNow = useSelector(state => state.events.activeNow);
    const {name} = useSelector(state => state.auth.account);

    const openModal = () => {
        dispatch(openModalEvents());
    }

    const btnDelete = () => {
        dispatch(deleteEvent(activeNow.id));
        dispatch(activeEvent(null));
    }

    const logout = ()=>{
        dispatch(startLogout());
    }

    return (
        <Box sx={styles.flexGrow_1} >
            <AppBar position="static" color='primary'>
                <Toolbar>

                    <Stack direction="row" spacing={2}>
                        <Avatar {...stringAvatar(`${name} +`)} />
                    </Stack>

                    <Typography variant="span" component="span" sx={styles.displayName}>
                        {name}
                    </Typography>

                    {
                        activeNow !== null &&
                        <Tooltip title="borrar evento">
                            <IconButton onClick={btnDelete} color='warning' aria-label="addEvent" component="span">
                                <DeleteForeverIcon />
                            </IconButton>
                        </Tooltip>
                    }

                    {
                        activeNow !== null ?
                        <Tooltip title="editar evento">
                            <IconButton onClick={openModal} color="inherit" aria-label="addEvent" component="span">
                                <EditIcon fontSize='small' />
                            </IconButton>
                        </Tooltip> :
                        <Tooltip title="agregar evento">
                            <IconButton onClick={openModal} color="inherit" aria-label="addEvent" component="span">
                                <EventIcon fontSize='small' />
                            </IconButton>
                        </Tooltip>
                    }

                    <Tooltip title="cerrar sesion">
                        <IconButton onClick={logout} color="inherit" aria-label="logout" component="span">
                            <LogoutIcon fontSize='small' />
                        </IconButton>
                    </Tooltip>

                </Toolbar>
            </AppBar>
        </Box>
    )
}
