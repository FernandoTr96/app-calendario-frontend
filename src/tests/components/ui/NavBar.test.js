import configure from 'redux-mock-store'
import thunk from 'redux-thunk'
import {mount} from 'enzyme'
import { NavBar } from '../../../components/ui/NavBar';
import { Provider } from 'react-redux';
import {deleteEvent, activeEvent} from '../../../actions/calendarActions'

jest.mock('../../../actions/calendarActions', ()=>({
    deleteEvent: jest.fn(),
    activeEvent: jest.fn()
}))

const middlewares = [thunk];
const mockStore = configure(middlewares);
const initState = {
    events: {
        activeNow: {
            id: '9376382223',
            title: 'prueba',
            desc: 'prueba',
            start: new Date().getTime(),
            end: new Date().getTime() + 1000
        }
    },
    auth:{
        account: {
            name: 'Fernando'
        }
    }
};
let store = mockStore(initState);
store.dispatch = jest.fn();

let wrapper = mount(
    <Provider store={store}>
        <NavBar/>
    </Provider>
)

describe('Pruebas en <NavBar/>', () => {

    beforeEach(()=>{
        wrapper = mount(
            <Provider store={store}>
                <NavBar/>
            </Provider>
        )
    })
    
    test('debe renderizarse correctamente', () => {
        
        expect(wrapper).toMatchSnapshot();

    });

    test('DeleteForeverIcon debe eliminar el evento', () => {

        wrapper.find('span').at(1).prop('onClick')();
        expect(deleteEvent).toHaveBeenCalledWith(initState.events.activeNow.id);

    });

});