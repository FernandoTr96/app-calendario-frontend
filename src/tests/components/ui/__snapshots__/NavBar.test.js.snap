// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <NavBar/> debe renderizarse correctamente 1`] = `
<div
  className="MuiBox-root css-i9gxme"
>
  <header
    className="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionStatic css-hip9hq-MuiPaper-root-MuiAppBar-root"
  >
    <div
      className="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-1t29gy6-MuiToolbar-root"
    >
      <div
        className="css-e53awj-MuiStack-root"
      >
        <div
          className="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-1vkgqbc-MuiAvatar-root"
        >
          F+
        </div>
      </div>
      <span
        className="MuiTypography-root MuiTypography-span css-1w0zw7x-MuiTypography-root"
      >
        Fernando
      </span>
      Array [
        <span
          aria-label="addEvent"
          aria-labelledby={null}
          className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorWarning MuiIconButton-sizeMedium css-cviz33-MuiButtonBase-root-MuiIconButton-root"
          data-mui-internal-clone-element={true}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseOver={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          role="button"
          tabIndex={0}
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="DeleteForeverIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zm2.46-7.12 1.41-1.41L12 12.59l2.12-2.12 1.41 1.41L13.41 14l2.12 2.12-1.41 1.41L12 15.41l-2.12 2.12-1.41-1.41L10.59 14l-2.13-2.12zM15.5 4l-1-1h-5l-1 1H5v2h14V4z"
            />
          </svg>
          <span
            className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </span>,
        "",
      ]
      Array [
        <span
          aria-label="addEvent"
          aria-labelledby={null}
          className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeMedium css-zylse7-MuiButtonBase-root-MuiIconButton-root"
          data-mui-internal-clone-element={true}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseOver={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          role="button"
          tabIndex={0}
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
            data-testid="EditIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 0 0-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
            />
          </svg>
          <span
            className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </span>,
        "",
      ]
      Array [
        <span
          aria-label="logout"
          aria-labelledby={null}
          className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeMedium css-zylse7-MuiButtonBase-root-MuiIconButton-root"
          data-mui-internal-clone-element={true}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseOver={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          role="button"
          tabIndex={0}
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
            data-testid="LogoutIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="m17 7-1.41 1.41L18.17 11H8v2h10.17l-2.58 2.58L17 17l5-5zM4 5h8V3H4c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h8v-2H4V5z"
            />
          </svg>
          <span
            className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </span>,
        "",
      ]
    </div>
  </header>
</div>
`;
